### ISSUE TITLE
使用mongodb和缓存获取数据错乱

### ISSUE THREAD

**Issue by themoonbear** [zeromicro/go-zero#2279]
![4B1FDC46-FCAE-4537-85CC-12ED1758A7A3](https://user-images.githubusercontent.com/2820281/185571120-f4a15d7f-bfd9-461c-aa1d-89aea7724f0d.png)
`

**Comment by kevwan** [comment]
Because too many requests to access the same non-exist data will bring down the service.

**Comment by themoonbear** [comment]
我的英文不太好, 还是使用中文表达吧. 从生成的代码看, 插入的时候,数据没有进入缓存, 获取的时候也没有从缓存中获取, 那这个缓存层在哪里起作用的呢? 有没有例子可以参考一下

**Comment by kevwan** [comment]
Delete when insert, and queries checks cache first.

**Comment by themoonbear** [comment]
> Delete when insert, and queries checks cache first.
看源码有看到, 在插入的时候删除了占位符的缓存, 但是每次调用FindOne的时候也是用的占位符, 那不是还是从数据库里取数据吗? 就是这块没看懂.

**Comment by kevwan** [comment]
One minute for placeholder cache.

**Comment by themoonbear** [comment]
在插入的时候 `if !data.ID.IsZero() {}` 这段代码是bug吧? 应该是空的时候才生成一个吧?

**Comment by kevwan** [comment]
Generated code?

**Comment by themoonbear** [comment]
> 

yes

**Comment by anqiansong** [comment]
感谢反馈，我尽快修复
> Thanks to your contribution, I'll fix it.

**Comment by anqiansong** [comment]
InsertOne 中的 key 是删除缓存，不管 id 是否存在，对业务都不影响，FindOne 不可能有空 ID，因为已经做了id 校验
> The `key` in function InsertOne is used to delete redis cache, it has no business effect whether the id exists or not, and the `key` in function FindOne, it is impossible that the case as you said it will return a empty data if the ID is empty, because there has a validation for ID before invoke `FindOne`.

**Comment by themoonbear** [comment]
> InsertOne 中的 key 是删除缓存，不管 id 是否存在，对业务都不影响，FindOne 不可能有空 ID，因为已经做了id 校验
> 
> > The `key` in function InsertOne is used to delete redis cache, it has no business effect whether the id exists or not, and the `key` in function FindOne, it is impossible that the case as you said it will return a empty data if the ID is empty, because there has a validation for ID before invoke `FindOne`.

你好，我理解设计的目的，但从用goctl生成的代码看，我不确定这样的代码逻辑是否正确，所有想确认一下是否是goctl生成的代码是错误的。
从生成代码看，FindOne使用的是空ID下去查的，你说了已做了ID校验请问是在哪一步？

**Comment by themoonbear** [comment]
> 

你好，我说的是截图里划红框的位置，缓存的KEY，截图代码是用goctl生成的，没有做过改动。在插入时判断插入数据是否有ID，如果有要重新生成一个，如果没有则缓存ID是空ID，执行过程中会删除空ID对应的缓存。在查询时，是用空ID到缓存里查，没有则从数据库里加到缓存中。这样这条数据是已空ID做为key存在缓存中的。

**Comment by kevwan** [comment]
Yes, it should be in cache. I've talked couple times in my presentations.

If we don't put the placeholder in cache, the cache cluster might be flooded for not-existing data.

**Comment by themoonbear** [comment]
> Yes, it should be in cache. I've talked couple times in my presentations.
> 
> If we don't put the placeholder in cache, the cache cluster might be flooded for not-existing data.

之前较忙没有继续跟进这个问题，占位符的设计目的无论从论坛和你的英文回复中我都清楚了，但并没有解决我遇到的问题，第一：调用InsertOne方法插入数据后如何知道该条记录在MongoDB中的ID值。第二：数据库中有两条数据，在调用FindeOne时由于第一条被占位符缓存，第二条取到的是第一条的数据。
![Dingtalk_20220930141503](https://user-images.githubusercontent.com/2820281/193203238-c5a9e0b8-60e1-4d50-8e56-910ef444cb4f.jpg)
![Dingtalk_20220930141651](https://user-images.githubusercontent.com/2820281/193203243-c2215d9d-069e-4647-b43c-492c4e5ff025.jpg)

**Comment by themoonbear** [comment]
@anqiansong 你好，可否跟进我的测试结果给与解答？
